### PR TITLE
[APE-1598] Add the management of range in where clause

### PR DESCRIPTION
**Pull request by JMGDoctolib**

### CODE REVIEW

File: lib/couchbase-orm/utilities/query_helper.rb [hunk 1]
```diff
@@ -87,6 +89,13 @@ def build_match_hash(key, value)
                 matches.join(" AND ")
             end
 
+            def build_match_range(key, value)
+                matches = []
+                matches << "#{key} >= #{quote(value.begin)}"
+                matches << "#{key} <= #{quote(value.end)}"
+                matches.join(" AND ")
```
Comment (JulienBURNET-FAUCHE) [inline lib/couchbase-orm/utilities/query_helper.rb]: Would it be the same as
 `"#{key} >= #{quote(value.begin)} AND #{key} <= #{quote(value.end)}"`

Comment (JMGDoctolib) [reply lib/couchbase-orm/utilities/query_helper.rb]: Yes it is. I found the other way easier to read but I can change if you prefer?

**Review by akermabon — COMMENTED** [review]
I think you should handle ranges with or without excluded end.
`1..3` => `key >= 1 and key <= 3`
`1...3` => `key >= 1 and key < 3`
You could rely on `Range` method `exclude_end?`

File: lib/couchbase-orm/utilities/query_helper.rb [hunk 2]
```diff
@@ -119,7 +132,7 @@ def serialize_value(key, value_before_type_cast)
             end
 
             def quote(value)
-                if value.is_a? String
+                if value.is_a? String or value.is_a? DateTime or value.is_a? Date or value.is_a? Time
```
Comment (akermabon) [inline lib/couchbase-orm/utilities/query_helper.rb]:
```suggestion
                if [String, DateTime, Date, Time].any? { |clazz| value.is_a?(clazz) }
```

**Review by akermabon — APPROVED** [review]
Looks good, one nitpick

File: lib/couchbase-orm/utilities/query_helper.rb [hunk 3]
```diff
@@ -18,6 +18,8 @@ def build_match(key, value)
                     "(#{build_match(key, nil)} OR #{build_match(key, value.compact)})"
                 when value.is_a?(Array)
                     "#{key} IN #{quote(value)}"
+                when value.instance_of?(Range)
```
Comment (pimpin) [inline lib/couchbase-orm/utilities/query_helper.rb]: I'd rather also accept any instance which class inherit from Range... or maybe any object responding to `:begin` and `:end`

```suggestion
                when value.is_a?(Range)
```

image

**Review by JulienBURNET-FAUCHE — APPROVED** [review]
Do we handle every _needs to be quoted_ cases here ? 
I am not sure, but at least, it's better than before

**Review by pimpin — APPROVED** [review]
LGTM.

**Comment by JMGDoctolib** [comment]
> Do we handle every _needs to be quoted_ cases here ? I am not sure, but at least, it's better than before

I don't know but at least I had the one we need.
Because today we can't do a `where(start_date: DateTime(...))`, it's failing 😢

**Pull request merged**